DOCS-5376 Update Nextjs Quickstart (#703)
clerk/clerk-docs: docs/quickstarts/nextjs.mdx

</CodeBlockTabs>

### Set your environment keys

1. If you don't have a `.env.local` file in the root directory of your Next.js project, create one now.

> 2. Find your secret keys. If you're signed into Clerk, the `.env.local` snippet below will contain your secret key. Otherwise:

Reviewer: Sorry, I didn't see this before, but this isn't just for their secret key. It's for their publishable key too.

</CodeBlockTabs>

### Set your environment keys

1. If you don't have a `.env.local` file in the root directory of your Next.js project, create one now.

> 2. Find your secret keys in your Clerk Dashboard:
>     - Navigate to your [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=api-keys) to find your secret keys.
>     - Select your application, then select **🔑 API Keys** in the sidebar menu.
>     - You can copy your keys from the **Quick Copy** section.
> 3. Add your keys to your `.env.local` file. The final result should look as follows:

Reviewer: Removing "to find your secret keys" as it seems redundant - we say this in the line above.

```suggestion
2. Find your secret keys in your Clerk Dashboard:
    - Navigate to your [Clerk Dashboard](https://dashboard.clerk.com/last-active?path=api-keys).
    - Select your application, then select **🔑 API Keys** in the sidebar menu.
    - You can copy your keys from the **Quick Copy** section.
3. Add your keys to your `.env.local` file. 

The final result should look as follows:

```

    <InjectKeys>

    ```sh filename=".env.local"
    NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY={{pub_key}}
    CLERK_SECRET_KEY={{secret}}
    ```

    </InjectKeys>

<Callout type="info">

If you're signed into Clerk, the above snippet will contain your secret key. Select the eye icon to reveal it. Otherwise, find your API keys in [the Clerk Dashboard](https://dashboard.clerk.com/last-active?path=api-keys).

> </Callout>

Reviewer: I think this can get moved to be above the "Inject keys" part. and it won't need to say "Otherwise, find your API keys..." any more because with your new additions, there are clear instructions of how to find them.

Author: Good point 👍 I've added this line to the instructions above the code snippet instead

```sh filename=".env.local"
NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY={{pub_key}}
CLERK_SECRET_KEY={{secret}}
```
<Callout type="info">

</InjectKeys>
If you're signed into Clerk, the above snippet will contain your secret key. Select the eye icon to reveal it. Otherwise, find your API keys in [the Clerk Dashboard](https://dashboard.clerk.com/last-active?path=api-keys).

This sets your public and secret keys.
</Callout>

### Wrap your app in `<ClerkProvider>`
### Add `<ClerkProvider>` to your app

The [`<ClerkProvider>`](/docs/components/clerk-provider) component provides active session and user context to Clerk's hooks and other components. Import it into your app by adding `import { ClerkProvider } from '@clerk/nextjs'` at the top of your file.
The [`<ClerkProvider>`](/docs/components/clerk-provider) component provides active session and user context to Clerk's hooks and other components.

Select your preferred router below to learn how to make this data available across your entire app:

<Tabs type="router" items={["App Router", "Pages Router"]}>
  <Tab>
    <Callout>
      The `app/layout.tsx` file might be in your `src` folder if it isn't in your root folder.
    </Callout>

>     For this project, it makes sense to wrap the `<html />` element with `<ClerkProvider>`. This makes the active session and user context accessible anywhere within the app.

Reviewer: I'm so glad you removed this

      children: React.ReactNode
    }) {
      return (
        // Active session and user context

>         // available throughout your app

Reviewer 2: These are the most common things that people pull from provider, but there are a lot of other things you can pull from it as well, so I'm not sure we'd want to add a comment like this rather than maybe like a link to clerk provider docs somewhere

Reviewer: yeah I agree, no need for the comment. the highlighting of the lines plus the copy is enough. 
and we have a link to clerkprovider in the copy above this code example

      children: React.ReactNode
    }) {
      return (

>         // Active session and user context

Author: ```suggestion
```

  </Tab>

  <Tab>
    ```tsx filename="_app.tsx" {2,8-10}
      import '@/styles/globals.css'
      import { ClerkProvider } from "@clerk/nextjs";
      import type { AppProps } from "next/app";
      function MyApp({ Component, pageProps }: AppProps) {
        return (
          // Active session and user context

>           // available to all children of MyApp

Reviewer: same comments as [the one above](https://github.com/clerk/clerk-docs/pull/703/files#r1488356090)

  </Tab>
</Tabs>

### Add authentication to your app

Now that Clerk is installed and mounted in your application, you can decide which pages are public and which should require authentication to access:

1. Create a `middleware.ts` file at the root of your project, or in your `src/` directory if you have one.
2. In your `middleware.ts`, export Clerk's [`authMiddleware`](https://clerk.com/docs/references/nextjs/auth-middleware) helper:

>     ```tsx filename="middleware.ts" {1,3}

Reviewer: the highlighted LOC isn't necessary here.

```suggestion
    ```tsx filename="middleware.ts"

```

    import { authMiddleware } from "@clerk/nextjs";

    export default authMiddleware({});
    ```

> 3. With `authMiddleware()`, authentication will be enabled by default on all routes that your Next.js middleware runs on. You can specify valid routes using Next.js's [`matcher`](https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher). Add the following code to your `middleware.ts` to protect your entire application:

Reviewer 2: ```suggestion
3. With `authMiddleware()`, authentication will be enabled by default on all routes that your Next.js middleware runs on. You can specify valid routes using Next.js' [`matcher`](https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher). Add the following code to your `middleware.ts` to protect your entire application:

```

> 3. With `authMiddleware()`, authentication will be enabled by default on all routes that your Next.js middleware runs on. You can specify valid routes using Next.js's [`matcher`](https://nextjs.org/docs/app/building-your-application/routing/middleware#matcher). Add the following code to your `middleware.ts` to protect your entire application:

Reviewer 3: I have another comment below, but with `authMiddleware()` we suggest matching all routes as its built with an opt-out approach. The key here is that routes are auth enabled and protected.

From there, the developer would use one of two things to make routes accessible:
1. `publicRoutes: []` - this leaves the route with auth information available, but means that users who are not signed in can access the route.
2. `ignoredRoutes: []` - there is no protection and no auth information. Removing a route from the matcher would be equal to an `ignoredRoute: []`.

Author: Thanks, I've updated this paragraph to reflect this now, and also added `publicRoutes` and `ignoredRoutes` to the code snippet to introduce the concept to readers

    ```tsx filename="middleware.ts" {8}
    import { authMiddleware } from "@clerk/nextjs";

    export default authMiddleware({});

    export const config = {
      // Protects all routes, including api/trpc.

>       // Edit this to allow other routes to be public.

Reviewer 3: With `authMiddleware()` the recommendation is to use `publicRoutes: []` to make routes public. `authMiddleware()` has both `ignoredRoutes: []` and `publicRoutes: []`. Removing a route from the Matcher would make it ignored and not public -- there would be no auth information available for that route. This will change with v5 soon, where the new Middleware `clerkMiddleware()` is opt-in.

Also, while the old message is a little unwieldy with the link in the comment, adding that link resulted in a solid decrease in middleware related questions. I would strongly suggest keeping some version of the following, with the link, in the sample code (and above `authMiddleware()` since it pertains to that.

```
// This example protects all routes including api/trpc routes
// Please edit this to allow other routes to be public as needed.
// See https://clerk.com/docs/references/nextjs/auth-middleware for more information about configuring your Middleware
```

Author: Thanks for pointing this out! I'll update the text and add the link back 👍

      matcher: ["/((?!.+\\.[\\w]+$|_next).*)", "/", "/(api|trpc)(.*)"],
    };
    ```

### Try accessing your app

Access your app to verify that authentication is enabled:

1. Run your project with the following terminal command from the root directory of your project:
    <CodeBlockTabs type="installer" options={["npm", "yarn", "pnpm"]}>
      ```bash filename="terminal"
      npm run dev
      ```

      ```bash filename="terminal"
      yarn dev
      ```

      ```bash filename="terminal"
      pnpm dev
      ```

    </CodeBlockTabs>
2. Visit [`http://localhost:3000`](http://localhost:3000) to access your app. The Middleware will redirect you to your Sign Up page, provided by Clerk's [Account Portal](/docs/account-portal/overview) feature.
3. Sign up to gain access to your application.

> 4. Visit your app again. You should be able to access it now.

Reviewer 2: ```suggestion
4. Log into your new account. You should be able to access it now.

```

### Embed the `<UserButton />`

[`<UserButton />`](/docs/components/user/user-button) is a [prebuilt component](/docs/components/overview) that allows users to:
  - Log in.
  - Manage their account information.

>   - Log out.

Reviewer 2: Making this a list feels awkward to me, I might just add it to the above sentence, comma-separated

Reviewer: I agree! lists are great, but not always necessary

> It must be a child of [`<ClerkProvider>`](/docs/components/clerk-provider).

Reviewer 3: This could be moved up to the `<ClerkProvider />` and changed to something more like 'all Clerk components, hooks and functions but be a child of `<ClerkProvider />`.

Use the following code to add a `<UserButton />` to your app: 

<CodeBlockTabs type="router" options={["App Router", "Pages Router"]}>
```tsx filename="app/page.tsx" {1,8}
import { UserButton } from "@clerk/nextjs";

export default function Home() {
  return (
    <div className="h-screen">
      // Use afterSignOutUrl to redirect

>       // users after signing out

Reviewer 2: This prop will be unnecessary in v5 which is slated to go out this/next week

Reviewer: also, these are the incorrect comments for JSX. if we want to keep these comments for the v4 docs, the proper comments use {/* */} formatting.

Reviewer 2: Great call, I missed that one 🤦‍♂️

</CodeBlockTabs>

Then, visit your app's homepage at [`localhost:3000`](http://localhost:3000) to see the `<UserButton />` in action. It should show your avatar from the account you signed in with.

### Sign out

Finally, use the `<UserButton />` by selecting your avatar, and select **Sign out**. You should be redirected to the url you specified with `afterSignOutUrl`.

</Steps>

## Next steps

To see an example application that demonstrates user authentication, and session and organization management, explore one of Clerk's Next.js demo repos:

- [Clerk + Next.js App Router Demo](https://github.com/clerk/clerk-nextjs-demo-app-router)
- [Clerk + Next.js Pages Router Demo](https://github.com/clerk/clerk-nextjs-demo-pages-router)

## More resources

<div className="container mx-auto my-4">
  <div className="grid grid-cols-1 gap-6 md:grid-cols-2">

    <Cards title="Deploy to Production" description="Learn how to deploy your Clerk app to production." link="/docs/deployments/overview" cta="Learn More" />

    <Cards title="Deploy to Vercel" description="Learn how to deploy your Clerk app to production on Vercel." link="/docs/deployments/deploy-to-vercel" cta="Learn More" />

> 

Reviewer: I'd even argue these would be the last steps in the line up. I feel that users will want to understand things like how to build a custom sign in page, how to read user data, how to use hooks, etc. before they would care about deploying to production or to vercel. 

what do you think?

Author: That makes sense!

</CodeBlockTabs>

### Sign out of your application

Visit your Next.js application at [`localhost:3000`](http://localhost:3000), click on your avatar, and select **Sign out**. You are now back where you started: a sign-in form.
Then, visit your app's homepage at [`localhost:3000`](http://localhost:3000) to see the `<UserButton />` in action. It should show your avatar from the account you signed in with.

> ### Deploy your application

Reviewer: yayyy, good riddance!

</Steps>

## Next steps

Now try deploying your app to production:

- See [Deploy to Vercel](/docs/deployments/deploy-to-vercel#deploying-a-clerk-app-to-vercel).

> - Or, if you're using another provider, see [Deploy to Production](/docs/deployments/overview).

Reviewer: imo, i wouldn't even add this as a "next step", i would maybe add it as a card under "More resources".

Reviewer 3: I haven't loved this as a last step. I don't think most devs would try deploying to production at that point. Having it as a card is probably a good play until the doc restructure takes shape.

Author: Great point! I've moved both these to cards under More Resources!
